lib/builder.dart

```diff
@@ -96,45 +96,70 @@ abstract class _MacroBuilder extends Builder {
-      var end = clazz.leftBracket.charOffset;
-      classBuffer.writeln(originalSource.substring(start, end + 1));
-
-      for (var checker in macros.keys) {
-        implementedDecls.addAll(await maybeApplyMacro(
-                checker,
-                macros[checker]!,
-                element,
-                resolver,
-                classBuffer,
-                libraryBuffer,
-                originalSource) ??
-            const []);
-      }
-      for (var field in element.fields) {
-        await _applyMacros(
-            field, classBuffer, libraryBuffer, resolver, originalSource,
-            implementedDecls: implementedDecls);
-      }
-      for (var method in element.methods) {
-        await _applyMacros(
-            method, classBuffer, libraryBuffer, resolver, originalSource,
-            implementedDecls: implementedDecls);
-      }
-      for (var constructor in element.constructors) {
-        await _applyMacros(
-            constructor, classBuffer, libraryBuffer, resolver, originalSource,
-            implementedDecls: implementedDecls);
-      }
-      for (var accessor in element.accessors) {
-        await _applyMacros(
-            accessor, classBuffer, libraryBuffer, resolver, originalSource,
-            implementedDecls: implementedDecls);
+      var clazz = (await resolver.astNodeFor(element, resolve: true));
+      if (clazz is analyzer.ClassOrMixinDeclaration) {
+        var start = clazz.offset;
+        var end = clazz.leftBracket.charOffset;
+        classBuffer.writeln(originalSource.substring(start, end + 1));
+
+        for (var checker in macros.keys) {
+          implementedDecls.addAll(await maybeApplyMacro(
+                  checker,
+                  macros[checker]!,
+                  element,
+                  resolver,
+                  classBuffer,
+                  libraryBuffer,
+                  originalSource) ??
+              const []);
+        }
+        for (var field in element.fields) {
+          await _applyMacros(
+              field, classBuffer, libraryBuffer, resolver, originalSource,
+              implementedDecls: implementedDecls);
+        }
+        for (var method in element.methods) {
+          await _applyMacros(
+              method, classBuffer, libraryBuffer, resolver, originalSource,
+              implementedDecls: implementedDecls);
+        }
+        for (var constructor in element.constructors) {
+          await _applyMacros(
+              constructor, classBuffer, libraryBuffer, resolver, originalSource,
+              implementedDecls: implementedDecls);
+        }
+        for (var accessor in element.accessors) {
+          await _applyMacros(
+              accessor, classBuffer, libraryBuffer, resolver, originalSource,
+              implementedDecls: implementedDecls);
+        }
+        classBuffer.writeln('}');
+        buffer.writeln(classBuffer);
+      } else if (clazz is analyzer.EnumDeclaration) {
```

Comment:
Is this actually implementing some level of support for enums? It doesn't look like it to me but just checking.

Fwiw we probably would be fine with adding that support, we just haven't put thought into it yet.

Reply:
It is really a minimal level of support. The reason I added it in was because the current implementation throws an exception if you have any enums in your code since it will try to cast an enum to a `ClassDeclaration` (since enums are `ClassElement`s). In fixing that error, I realized how easy it would be to add in basic support for enums. Obviously a more specific macro would work better with enums, but I think ClassDeclarationMacro and FieldDeclarationMacro might work on enums as is. I haven't really tested macro support for enums except to verify that it doesn't rip out the enum definition and builds everything else fine.

Reply:
Ok, I think we will probably want a more specific thing for enums, but I will file an issue about it and this is fine for now, its just a prototype after all :).